android/quest/src/main/java/org/smartregister/fhircore/quest/ui/report/measure/MeasureReportViewModel.kt

```diff
@@ -335,28 +350,36 @@ constructor(
     }
   }
 
+  private suspend fun fetchSubjects(config: MeasureReportConfig): List<String> {
+    return reportTypeSelectorUiState.value.subjectViewData
+      .ifEmpty {
+        if (config.subjectXFhirQuery?.isNotEmpty() == true) {
+          fhirEngine.search(config.subjectXFhirQuery!!).map {
+            // TODO a hack to prevent missing subject in case of Group based reports where
+            // MeasureEvaluator looks for Group members and skips the Group itself
+            if (it is Group && !it.hasMember()) {
+              it.addMember(Group.GroupMemberComponent(it.asReference()))
+              fhirEngine.update(it)
+            }
+            MeasureReportSubjectViewData(it.resourceType, it.logicalId, it.valueToString())
+          }
+        } else emptyList()
+      }
+      .map { "${it.type.name}/${it.logicalId}" }
+  }
+
   private suspend fun evaluatePopulationMeasure(
     measureUrl: String,
     startDateFormatted: String,
     endDateFormatted: String,
-    subjectXFhirQuery: String?,
+    subjects: List<String>,
     existing: List<MeasureReport>
   ): List<MeasureReport> {
     val measureReport = mutableListOf<MeasureReport>()
     withContext(dispatcherProvider.io()) {
-      if (subjectXFhirQuery?.isNotEmpty() == true) {
-        fhirEngine
-          .search(subjectXFhirQuery)
-          .map {
-            val subject = "${it.resourceType}/${it.logicalId}"
-            // TODO a hack to prevent missing subject in case of Group based reports where
-            // MeasureEvaluator looks for Group members and skips the Group itself
-            if (it is Group && !it.hasMember()) {
-              it.addMember(Group.GroupMemberComponent(it.asReference()))
-              fhirEngine.update(it)
-            }
-            runMeasureReport(measureUrl, SUBJECT, startDateFormatted, endDateFormatted, subject)
-          }
+      if (subjects.isNotEmpty()) {
+        subjects
+          .map { runMeasureReport(measureUrl, SUBJECT, startDateFormatted, endDateFormatted, it) }
```

Comment:
this looks good, at a later date we can experiment with using mapp here, but totally possible that running in parallel might actually lead to worse performance

Reply:
Yeah I think we can try this and see